test/common/upstream/health_checker_impl_test.cc

```diff
@@ -4529,21 +4529,176 @@ TEST_F(GrpcHealthCheckerImplTest, GrpcFailUnknownHealthStatus) {
             cluster_->prioritySet().getMockHostSet(0)->hosts_[0]->health());
 }
 
-// Test receiving GOAWAY is interpreted as connection close event.
+// Test receiving GOAWAY is handled gracefully while a check is in progress.
```

Comment:
This tells me that the code intentionally treated GOAWAY as failures, which makes me wonder if this is the desired behavior for some? Does this mean that this should be configurable?

@baranov1ch @htuch @mattklein123 WDYT

Reply:
This might depend on the reason for the GOAWAY? I.e. a `NO_ERROR` mid check might be fine as it's telling to drain, but not if it was an error.

Reply:
I see, that makes sense. 

It looks like the GOAWAY error code isn't provided in the callback at the moment, but I think it would be reasonable to do that?

It seems that the `onGoAway()` callbacks are only fired once (see [here](https://github.com/envoyproxy/envoy/blob/f71ab4cfb1a7293f2ffb3a7271550d719ee01393/source/common/http/http2/codec_impl.cc#L580-L587)), and - with error codes - we'd want to call them for each GOAWAY frame received. I'm not sure if that would be problematic. For reference it looks the "fire once" behaviour was added in https://github.com/envoyproxy/envoy/pull/103.

It may also be worth mentioning that the http2 connection pool implementation doesn't change behaviour based on the GOAWAY error code at the moment (see [here](https://github.com/envoyproxy/envoy/blob/f71ab4cfb1a7293f2ffb3a7271550d719ee01393/source/common/http/http2/conn_pool.cc#L27-L37)). It transitions the active client to the draining state if it has active requests.

Reply:
Right, if the GOAWAY error code is NO_ERROR we should just treat it as "don't reuse connection", not an error. We may have to add some additional error code info to the callback. Note that @antoniovicente just opened https://github.com/envoyproxy/envoy/issues/11420, though IIRC nghttp2 does provide the reason code inside the frame here: https://github.com/envoyproxy/envoy/blob/5d8f7d785b59d1450b67f3279706b787c5641a66/source/common/http/http2/codec_impl.cc#L582, we just need to store it.

Reply:
> IRC nghttp2 does provide the reason code inside the frame here:

Yep - it looks like it's provided in `frame->goaway.error_code` and the debug data in `frame->goaway.opaque_data`


I'm happy to update the `onGoAway()` interface as part of this PR if @antoniovicente doesn't already have something started.

Does something along these lines look right? 
```
  # Maybe pass an Envoy-internal enum rather than the nghttp2 enum?
  virtual void onGoAway(nghttp2_error_code error_code, absl::Span<uint8_t> debug_data) PURE;
```

Reply:
We should not leak nghttp2 specific structs to the public interface, so at least for the error code we should do a mapping. For your use case maybe just have an enum with {NO_ERROR, OTHER} and @antoniovicente can fill out more stuff later as needed?

Reply:
enum with {NO_ERROR, OTHER} sounds like a good start